content/articles/mocking-with-junit-and-mockito-the-why-and-how/index.md

```diff
@@ -0,0 +1,871 @@
+
+> Notice that we have removed the ```when -then``` directives from our test case.
+We refactor the test case for the ```registerStudent``` method as follows:
+
+```java
+ @Test
+    void registerStudentTest() throws Exception {
+      RegistrationRequest registrationRequest = new RegistrationRequest("John", "Fallow",
+      "my_password", "PSC1004396", Gender.MALE);
+      Student student = Student.builder()
+                    .firstName("John")
+                    .lastName("Fallow")
+                    .matricNo("PSC100419")
+                    .password("my_password")
+                    .gender(Gender.MALE).build();
+      doReturn(student).when(studentRepository).save(any(Student.class));
+      StudentDto studentDto = studentService.registerStudent(registrationRequest);
+      assertThat(studentDto, hasProperty("firstName", equalTo("John")));
+      assertThat(studentDto, hasProperty("lastName", equalTo("Fallow")));
+      assertThat(studentDto, hasProperty("matricNo", equalTo("PSC100419")));
+      assertThat(studentDto, hasProperty("gender", equalTo(Gender.MALE)));
+  }
+```
+
+> Notice ```doReturn(student).when(studentRepository).save(any(Student.class));``` is similar to writing ```when(studentRepository.save(any(Student.class)).thenReturn(student)```.  There is an important pitfall to note when using spy objects with the  ```when then```directive. The actual method will be called (because ```studentRepository.save(any(Student.class))``` is evaluated before ```when(..)``` at runtime). Using the ```when- then``` directive with spies can be a problem if our method contains logic that we do not want to invoke. Hence, we use```doReturn(student).when(studentRepository).save(any(Student.class));```
+
+At this point, it is important to highlight that the ```do…when``` directive can be used as an alternative to the ```when -then``` directive:
+- ```when(studentRepository.save(any(Student.class))).thenReturn(student)``` can be rewritten as ```doReturn(student).when(studentRepository).save(any(Student.class))```
+
+- ```when(studentRepository.save(null)).thenThrow(NullEntityException.class)``` can be rewritten as ```doThrow(NullEntityException.class).when(studentRepository).save(null)```
+
+- ``` java
+  when(studentRepository.save(any(Student.class))).thenAnswer(answer->{
+  return Student.builder()
+  .firstName("John")
+  .lastName("Doe")
+  .matricNo("MAT100419")
+  .password("securedPassword")
+  .registrationTime(getTime())
+  .gender(Gender.MALE)
+  .build();
+  });
+  ```
+  can be rewritten as:
+  ```java
+   doAnswer(answer->{ return Student.builder()
+                   .firstName("John")
+                   .lastName("Doe")
+                   .matricNo("MAT100419")
+                   .password("securedPassword")
+                   .registrationTime(getTime())
+                   .gender(Gender.MALE).build();
+      }).when(studentRepository).save(any(Student.class))```
```

Comment:
You seem to have added this backticks by mistakes, could you remove them

Reply:
@ehizman, I have managed to run all the test cases in your article and they all passed 👍🏾 . Kindly fix those minor issues that I have highlighted and we will be done with tech review

Reply:
Done